### ISSUE TITLE
Map node lang/late to country name

### ISSUE THREAD

**Issue by 0oM4R** [threefoldtech/tfchain_graphql#190]
Map the country names in GraphQL:
on create node, GraphQL should use the longitude and latitude to set the country name using another service, so we will ignore the country and city fields from  the chain and will overwrite it, and we can also add country code to manage regions in gridproxy

_Originally posted by @0oM4R in https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3021#issuecomment-2282836629_

we should use third party package to determine the country code then map it to the county table and update node location with the new values

**Comment by 0oM4R** [comment]
Issue Update: 
I was searching for a tool to get the location data using longitude and latitude, i couldn't find any npm package that match our requirement. 
but I have multiple options: 
1- using https://geopy.readthedocs.io/en/stable/#installation
2 using OSM api `https://nominatim.openstreetmap.org/reverse.php?lat=31.0371&lon=32.3844&zoom=18&format=jsonv2`
3- i tried to check the maxmind database, We can get the `geoname_id` using lon/lat then get location data using that id, or we may manipulate the database and  update the current country table to contain lat/long and its location
![image](https://github.com/user-attachments/assets/f95c3ac8-0145-401f-9f12-480dd866b283)


I don't think it is a good idea to rely on api or third party service, we may reach to another conflict later, i think we should maintain our database, and we can add it to geoip.grid.tf as a service

**Comment by 0oM4R** [comment]
Issue update:
I wrote a script that merge ipv4 and ipv6 data and dropped the unnecessary columns,
working on cleaning the data and removing redundant coordinates

**Comment by 0oM4R** [comment]
Blocked on https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3021#issuecomment-2295200776

**Comment by 0oM4R** [comment]
issue update: will catch the update event and update the location on graphql, 
will give it 3 hours to investigate

**Comment by 0oM4R** [comment]
Location details got updated on `UpdateNode` event

**Comment by xmonader** [comment]
awaiting https://github.com/threefoldtech/tfgrid-sdk-go/pull/1260

**Comment by scottyeager** [comment]
https://github.com/threefoldtech/tfgrid-sdk-go/pull/1260 was merged on March 3. Any further holdup here?

**Comment by sameh-farouk** [comment]
### Update (Apr 2026)

#### Context

The on-chain `country` field is a free-text string set by ZOS via `geoip.grid.tf`. Different nodes have different naming for the same country because:

1. **The geoip database changed names over time**: "Netherlands" → "The Netherlands", "Turkey" → "Türkiye"
2. **The processor's country reference table** (`countries.json` / `init-countries.js`) uses a different naming convention than geoip
3. **Non-production networks have fake registrations** with arbitrary strings
4. **Offline/dead nodes never self-correct**

This originally broke grid-proxy's `JOIN node.country = country.name` (used for region filtering) and dashboard country-based filtering.

#### How location flows through the system

```
ZOS boot / periodic updater (every 24h since Sep 2024)
  → geoip.Fetch() → geoip.grid.tf
  → gets {Country, City, Lat, Lon, CountryCode, Continent}
  → submits Country, City, Lat, Lon to chain (CountryCode and Continent are DISCARDED)
  → chain emits NodeStored / NodeUpdated
  → processor stores raw country string
```

Key finding: **ZOS already has `CountryCode` (ISO 3166-1 alpha-2) from geoip but never submits it to chain.** The `LocationInput` struct on-chain only has `Country`, `City`, `Latitude`, `Longitude`.

#### Prior work done

| Date | What | Issue/PR |
|------|------|----------|
| Feb 2024 | Switched `init-countries.js` from restcountries v2 (official names) to v3 (common names). Fixed Czechia/US/UK mismatch. | PR #158, fixes #148 |
| Jun 2024 | Bundled `countries.json` locally after `restcountries.com` went down during deployment. Added `ON CONFLICT DO UPDATE`. | PR #185, fixes #181 |
| Sep 2024 | ZOS periodic location updater. Re-fetches geoip every 24h, updates on-chain if changed. | zosbase commit `435ca91f` |
| Mar 2025 | Grid-proxy location indexer. Calls live nodes via RMB for fresh geoip data, stores country + continent in `node_location` table. Replaced the old `country` reference table join. | tfgrid-sdk-go PR #1260, fixes #1248 |
| Jun 2025 | Grid-proxy adds region field to node response. | tfgrid-sdk-go PR #1367, refs #1366 |

#### ZOS self-correction behavior

- **On boot:** `identityd` fetches fresh geoip and calls `UpdateNode` if location data changed. This was always the case.
- **Periodic updater (since Sep 2024, commit `435ca91f`):** A location updater was added to ZOS that re-fetches geoip every 24 hours while running and calls `UpdateNode` if data changed. Live nodes on a recent ZOS version self-correct within 24 hours of a geoip naming change, no reboot needed.
- **Nodes that remain inconsistent:** offline/dead nodes, and nodes running ZOS from before Sep 2024 that haven't received an upgrade.

#### Grid-proxy location indexer: what it actually fixes

The grid-proxy location indexer (PR #1260) replaced the old broken `JOIN node.country = country.name` (against the processor's country reference table) with `JOIN node.country = node_location.country` (against its own table). This join works because both sides come from the same geoip source: the node writes its country to chain, and the indexer calls the same node's geoip via RMB and gets the same value. So the join matches regardless of naming convention.

The indexer also adds **continent** data, which the chain and processor don't have. Note that the country reference table has a `region` field ("Americas", "Africa", "Europe", etc.) but these don't match MaxMind continent names exactly ("North America"/"South America" vs "Americas", "Antarctica" vs "Antarctic").

**Grid-proxy's naming problem is effectively solved** by PR #1260 for live nodes. The remaining issue is on the GraphQL side, where the processor's `Country` entity (from `countries.json`) still has naming that doesn't match some nodes. This only affects direct GraphQL API consumers querying the Country entity.

#### Why name-based matching didn't/won't work

Fuzzy or contains matching is unreliable. "Guinea" matches "Papua New Guinea", "Equatorial Guinea", and "Guinea-Bissau". An aliases map is whack-a-mole that breaks whenever geoip updates another name.

The only reliable standardization key is the **ISO country code**. ISO codes change only for major geopolitical events (handful per decade), unlike free-text names that change based on DB vendor decisions.

#### Two paths forward (if we really still need to fix this in GraphQL layer)

**Option 1: Chain change (recommended)**

Add `countryCode` to `LocationInput` on-chain. ZOS already has it from geoip, just needs to submit it. Processor stores it as the canonical key. This is the most reliable solution because the code comes from the same geoip source as the coordinates, guaranteed to be consistent.

Requires a spec upgrade + ZOS update + processor schema change.

**Option 2: Processor-side boundary lookup (fallback, best-effort)**

On `nodeStored`/`nodeUpdated`, use the node's lat/lon with a country boundary dataset (e.g., `which-country` npm package) to determine the ISO code. However this has issues:

- **We're essentially re-deriving what geoip already computed.** Geoip returns the correct `CountryCode` from its own database, but the chain discards it. Using boundary lookup on geoip's coordinates adds a layer of inaccuracy on top of data that was already correct at the source, since the boundary dataset may disagree with geoip on edge cases.
- Nodes with fake/zero coordinates (common on non-prod networks) could produce wrong or no match
- Coordinates in the ocean (geoip sometimes returns country-center when city is unknown) produce no match
- Border precision issues could resolve to wrong country
- Disputed territories are dataset-dependent

Could work as best-effort (~99% accuracy) with a safety rule: use boundary result if matched, keep raw chain value otherwise. But it's a strictly inferior version of Option 1.

#### Additional findings

- `init-countries.js` still fetches `countries.json` from GitHub raw URL instead of reading the local bundled file (PR #185 bundled the file but didn't change the fetch URL)
- Cities still come from external API (`shivammathur/countrycity`)
- The country table already has ISO codes (`cca2`), regions, subregions. The normalization reference exists
- tfgrid-sdk-go#1366 (add region to node response) has PR #1367 merged but the issue appears still in verification

#### Final note and recommendation

Besides the two paths mentioned above, "doing nothing" is a valid option if this is low priority with the focus on the Hero Ledger migration. GraphQL is successfully indexing TFChain and as long as its data matches the on-chain data (the source of truth), it is functioning exactly as intended. The Country reference table (from countries.json) is an auxiliary dataset loaded by a helper script, not chain-derived data.

The naming mismatch between this table and geoip-sourced node country names is now low impact since grid-proxy (the main consumer) solved it with the location indexer. We could close this as won't-fix if that's acceptable.

**Comment by sameh-farouk** [comment]
Closing this as good enough for now
